src/pkcs11/pkcs11-global.c

```diff
@@ -656,9 +656,13 @@ CK_RV C_WaitForSlotEvent(CK_FLAGS flags,   /* blocking/nonblocking flag */
 		mask |= SC_EVENT_READER_EVENTS;
 	}
 
+
+	sc_ctx_detect_readers(context);
+
 	rv = slot_find_changed(&slot_id, mask);
-	if ((rv == CKR_OK) || (flags & CKF_DONT_BLOCK))
+	if (rv == CKR_OK   ||  (rv != CKR_TOKEN_NOT_PRESENT  && (flags & CKF_DONT_BLOCK) ))
```

Comment:
I changed the return of carddetectAll in slot.c
On 8 Dec 2015 17:23, "Frank Morgner" <notifications@github.com> wrote:

> In src/pkcs11/pkcs11-global.c
> <https://github.com/OpenSC/OpenSC/pull/630#discussion_r46954556>:
>
> >  	rv = slot_find_changed(&slot_id, mask);
> > -	if ((rv == CKR_OK) || (flags & CKF_DONT_BLOCK))
> > +	if (rv == CKR_OK   ||  (rv != CKR_TOKEN_NOT_PRESENT  && (flags & CKF_DONT_BLOCK) ))
>
> this doesn't work. card_detect_all will never return CKR_TOKEN_NOT_PRESENT
> !
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenSC/OpenSC/pull/630/files#r46954556>.
>